### PR TITLE
tests: Bluetooth: Mesh: Proxy mix-in test

### PR DESCRIPTION
**Pull request by ardo-nordic**
When proxy or PB-GATT is enabled, the advertiser will transmit connectable advertisements whenever there are no mesh packets to be sent. As soon as a mesh packet enters the queue, connectable advertising will be halted, and the mesh packet is transmitted. Connectable advertisements are resumed as soon as the advertiser queue is empty again.

Feature coverage:
    PB-GATT advertisements
    GATT Proxy advertisements
    Mix of ADV and GATT advertisements

Test:

    PB-GATT and GATT Proxy must be enabled in the configuration.
    Initialize the mesh stack, but do not provision the device.
    Start the provisioning procedure using bt_mesh_prov_enable(BT_MESH_PROV_GATT)
    With a separate scanner device, observe PB-GATT beacons being sent at approximately 100ms intervals.

Provision the device using bt_mesh_provision().

    With a separate scanner device, observe GATT Proxy beacons being sent at approximately 1s intervals.
    Allocate an advertiser buffer using bt_mesh_adv_buf_create()
        With the separate scanner, observe that the GATT Proxy beacons are not interrupted.
    Set the allocated buffer's xmit parameters to BT_MESH_TRANSMIT(5, 20)
    Submit the advertiser buffer using bt_mesh_adv_send()
        With the separate scanner, observe that the submitted buffer is transmitted at least 6 times.
        Then, observe GATT Proxy beacons resuming.

### CODE REVIEW

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c
Comment (Balaklaka) [inline tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:197]: Is delta really needed here. It is zero.

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c
Comment (Balaklaka) [inline tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:224]: It is not wrong to have the k_sem_reset, but I believe it is not really necessary.

**Review by Balaklaka — CHANGES_REQUESTED** [review]

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c
Comment (alxelax) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:293]: It is set via this definition for full test suite: https://github.com/ardo-nordic/zephyr/blob/main/tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c#L20

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c
Comment (alxelax) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:401]: 100 msec is correct. 1800 msec is set to get a couple of pb-gatt beacons and collect several deltas between them to be sure they keep about 100ms interval. 
The same for proxy beacons.

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c
Comment (Andrewpini) [inline tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:419]: I guess the sleep sequences here for the most part is put in to sync with the RX counterpart? Should we consider merging the sync lib from mesh-1.1 branch to zephyr? IMO using ad-hoc sleep cycles to make the test work is error proned. Im not necessarily saying we must do this before merging this, but it should be considered.

**Review by Andrewpini — CHANGES_REQUESTED** [review]

File: tests/bluetooth/bsim_bt/bsim_test_mesh/prj_gatt.conf
Comment (alxelax) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/prj_gatt.conf:1]: I see the others do not have a license header. I guess this is a common approach for *.prj files in Zephyr.

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c
Comment (alxelax) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:461]: No idea. @PavelVPV has investigated a bit bugs related to ext advertiser and gatt bearer. Is this what you have seen? 
Any case this issue should be investigated separately not as part of test implementation.

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c
Comment (alxelax) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:419]: No, it affects significantly. The main purpose of it is to check the mix of transmitting via advertiser and GATT bearers. After sending data via the advertiser bearer above the test gives time to restore proxy beaconing via GATT bearer. Test gives time 6 seconds to send proxy beacons and rx part of test checks that.

Comment (alxelax) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:419]: The test is synced in a couple of places automatically using K_FOREVER semaphores. If rx part has been completed something faster it will wait for tx part in semaphore implicitly. I guess it removes potential instability.

Comment (alxelax) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:419]: btw. the test works for both extended advertiser and legacy advertiser for different PCs. Obviously, advertisers have different timelines.

Comment (Andrewpini) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:419]: Ok, but as far as I know passing the test for the tx_dev does not suspend it? It will still be runing until the RX dev either fails or passes?

File: tests/bluetooth/bsim_bt/bsim_test_mesh/prj_gatt.conf
Comment (alxelax) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/prj_gatt.conf:1]: I guess this is used for any mesh test suite:
https://github.com/ardo-nordic/zephyr/blob/main/tests/bluetooth/bsim_bt/bsim_test_mesh/prj.conf#L1
https://github.com/ardo-nordic/zephyr/blob/main/tests/bluetooth/bsim_bt/bsim_test_mesh/prj_low_lat.conf#L3
https://github.com/ardo-nordic/zephyr/blob/main/tests/bluetooth/bsim_bt/bsim_test_mesh/prj_pst.conf#L7

Which options should be removed you think?

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c
Comment (alxelax) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:419]: I'm not sure. This behavior is not documented anywhere. Seems it is relying on some internal bsim engine behavior. I'd prefer to keep the expected behavior within the test to be sure it is not changed after some modifications within BabbleSim (for instance after supporting a new chip).

Comment (Andrewpini) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:419]: I've tested it and it does run after passing the test. Does it not seem unlikely that the maintainers of bsim will alter this behavior? Anyways, if you really would prefer to keep this as is it is fine by me 👍

File: tests/bluetooth/bsim_bt/bsim_test_mesh/CMakeLists.txt
Comment (alxelax) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/CMakeLists.txt:26]: I guess it is better to solve issue if it happens. Right now we have only these combinations.

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c
Comment (alxelax) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c]: removed

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c
Comment (alxelax) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c]: made semaphore with 20 seconds waiting time

File: tests/bluetooth/bsim_bt/bsim_test_mesh/prj_gatt.conf
Comment (alxelax) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/prj_gatt.conf:1]: removed friend, lpn and IV index test features from configuration. If you think it is possible to remove more please point to exact options.

Comment (alxelax) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/prj_gatt.conf:1]: This is strange, because tx part performs self provisioning according to test scenario. Otherwise, it is not possible to start gatt proxy. :)

Comment (Andrewpini) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/prj_gatt.conf:1]: Ok, but it's no big deal. My point was that we should try to avoid cpy/paste for config files, since it is likely to add unecessary options.

**Review by Andrewpini — DISMISSED** [review]

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c
Comment (PavelVPV) [inline tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:458]: I guess this should be debugged.

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c
Comment (PavelVPV) [inline tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:406]: The comments could be a bit more meaningful. Comments rephrasing a function name aren't really helpful. E.g. here it could be something like:
```
/* Send a mesh message while advertising proxy service. Advertising the proxy service should be resumed after finishing advertising the message. */
```

**Review by PavelVPV — CHANGES_REQUESTED** [review]

**Review by Balaklaka — DISMISSED** [review]

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c
Comment (alxelax) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:431]: It should be separate pointer since it is used in scanner callback. No way to pass it there other than global value. It doesn't matter this is pointer or just variable. I don't think it has any sense, advantage or disadvantage to change it on variable or make it as incoming parameter for `rx_gatt_beacons`.

Which more meaningful name do you assume? When I gave name I assumed gatt session parameter.

Comment (PavelVPV) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:431]: Then either of this is possible and should make the code more readable:
- Access `gatt_param` directly in the test function. This is kind of already done in few lines below, but then I don't see why it can't be done initially.
- Instantiate `gattparam` in the test function and pass it by value or pointer to `rx_gatt_beacons`. Then it creates a clear connection between this structure and the function, saves few minutes of developers figuring out the relation of this code and makes life simpler;

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c
Comment (PavelVPV) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:422]: Then it should be clarified in the code, e.g.:
```suggestion
		.transmits = 1500 /* total adv duration */ / 100 /* interval*/,
```

This could even be macro taking number of advertisements and then used in both places, here and in the dut test function.

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c
Comment (alxelax) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:258]: due to controller tolerance. with increasing number of retransmits some of them cannot fit into previous range.

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c
Comment (alxelax) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:440]: with growing of interval transmit tolerance is decreased more than twice

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c
Comment (alxelax) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c:458]: yep but not as part of this PR

**Review by Andrewpini — DISMISSED** [review]

**Review by PavelVPV — CHANGES_REQUESTED** [review]
Can the first 3 commits be merged together? It would be good to explain in commit message why a separate test suite is used for GATT.

**Comment by alxelax** [comment]
> Can the first 3 commits be merged together? It would be good to explain in commit message why a separate test suite is used for GATT.

I do not think the commit with gatt test suite belongs logically to the first two commits.
I'll merge the first 2 but the third one should be separate.

**Comment by PavelVPV** [comment]
> > Can the first 3 commits be merged together? It would be good to explain in commit message why a separate test suite is used for GATT.
> 
> I do not think the commit with gatt test suite belongs logically to the first two commits. I'll merge the first 2 but the third one should be separate.

I guess this will brake one of [this](https://docs.zephyrproject.org/latest/contribute/guidelines.html#other-commit-expectations) rules:
```
Commits must build cleanly when applied on top of each other, thus avoiding breaking bisectability.
```

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c
Comment (alxelax) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c]: This is log of test with minor modification. I added logging of adv type in scanner. Scanner works as sniffer in this test. As you can see there is no another advertiser activity. Tolerance is the same as for adv bearer. I'll change in the test. Nevertheless, you can see that precision for pb-gatt advertising is higher than for proxy beaconing or for advertiser bearer.
```
Starting adv_tx_proxy_mixin as device #0
Starting adv_rx_proxy_mixin as device #1
Starting phy with 2 devices
d_01: @00:00:00.000000  *** Booting Zephyr OS build zephyr-v3.1.0-5224-ge2f00abe9248  ***
d_00: @00:00:00.000000  *** Booting Zephyr OS build zephyr-v3.1.0-5224-ge2f00abe9248  ***
d_01: @00:00:00.005768  [00:00:00.005,767] <inf> bt_hci_core: HW Platform: Nordic Semiconductor (0x0002)
d_01: @00:00:00.005768  [00:00:00.005,767] <inf> bt_hci_core: HW Variant: nRF52x (0x0002)
d_01: @00:00:00.005768  [00:00:00.005,767] <inf> bt_hci_core: Firmware: Standard Bluetooth controller (0x00) Version 3.2 Build 0
d_01: @00:00:00.005768  [00:00:00.005,767] <wrn> bt_id: No static addresses stored in controller
d_01: @00:00:00.006488  [00:00:00.006,469] <inf> bt_hci_core: Identity: FF:21:3D:E5:B4:CC (random)
d_01: @00:00:00.006488  [00:00:00.006,469] <inf> bt_hci_core: HCI: version 5.3 (0x0c) revision 0x0000, manufacturer 0x0059
d_01: @00:00:00.006488  [00:00:00.006,469] <inf> bt_hci_core: LMP: version 5.3 (0x0c) subver 0xffff
d_01: @00:00:00.006488  [00:00:00.006,469] <inf> test_adv: Bluetooth initialized
d_00: @00:00:00.005768  [00:00:00.005,767] <inf> bt_hci_core: HW Platform: Nordic Semiconductor (0x0002)
d_00: @00:00:00.005768  [00:00:00.005,767] <inf> bt_hci_core: HW Variant: nRF52x (0x0002)
d_00: @00:00:00.005768  [00:00:00.005,767] <inf> bt_hci_core: Firmware: Standard Bluetooth controller (0x00) Version 3.2 Build 0
d_00: @00:00:00.005768  [00:00:00.005,767] <wrn> bt_id: No static addresses stored in controller
d_00: @00:00:00.006488  [00:00:00.006,469] <inf> bt_hci_core: Identity: C5:AD:ED:38:00:B2 (random)
d_00: @00:00:00.006488  [00:00:00.006,469] <inf> bt_hci_core: HCI: version 5.3 (0x0c) revision 0x0000, manufacturer 0x0059
d_00: @00:00:00.006488  [00:00:00.006,469] <inf> bt_hci_core: LMP: version 5.3 (0x0c) subver 0xffff
d_00: @00:00:00.006488  [00:00:00.006,469] <inf> mesh_test: Bluetooth initialized
d_00: @00:00:00.006488  [00:00:00.006,469] <inf> mesh_test: Mesh initialized
d_00: @00:00:00.006488  [00:00:00.006,469] <inf> bt_mesh_prov_device: Device UUID: 00000000-0000-0000-0000-aa61676e696c
d_01: @00:00:00.023147  [00:00:00.023,132] <inf> test_adv: rx: adv type 0
d_01: @00:00:00.023147  [00:00:00.023,132] <inf> test_adv: rx: adv test +0 ms
d_01: @00:00:00.127364  [00:00:00.127,349] <inf> test_adv: rx: adv type 0
d_01: @00:00:00.127364  [00:00:00.127,349] <inf> test_adv: rx: adv test +107 ms
d_01: @00:00:00.234633  [00:00:00.234,619] <inf> test_adv: rx: adv type 0
d_01: @00:00:00.234633  [00:00:00.234,619] <inf> test_adv: rx: adv test +110 ms
d_01: @00:00:00.344344  [00:00:00.344,329] <inf> test_adv: rx: adv type 0
d_01: @00:00:00.344344  [00:00:00.344,329] <inf> test_adv: rx: adv test +112 ms
d_01: @00:00:00.449202  [00:00:00.449,188] <inf> test_adv: rx: adv type 0
d_01: @00:00:00.449202  [00:00:00.449,188] <inf> test_adv: rx: adv test +107 ms
d_01: @00:00:00.553725  [00:00:00.553,710] <inf> test_adv: rx: adv type 0
d_01: @00:00:00.553725  [00:00:00.553,710] <inf> test_adv: rx: adv test +108 ms
d_01: @00:00:00.654098  [00:00:00.654,083] <inf> test_adv: rx: adv type 0
d_01: @00:00:00.654098  [00:00:00.654,083] <inf> test_adv: rx: adv test +102 ms
d_01: @00:00:00.763930  [00:00:00.763,916] <inf> test_adv: rx: adv type 0
d_01: @00:00:00.763930  [00:00:00.763,916] <inf> test_adv: rx: adv test +113 ms
d_01: @00:00:00.870772  [00:00:00.870,758] <inf> test_adv: rx: adv type 0
d_01: @00:00:00.870772  [00:00:00.870,758] <inf> test_adv: rx: adv test +109 ms
d_01: @00:00:00.977828  [00:00:00.977,813] <inf> test_adv: rx: adv type 0
d_01: @00:00:00.977828  [00:00:00.977,813] <inf> test_adv: rx: adv test +110 ms
d_01: @00:00:01.087228  [00:00:01.087,219] <inf> test_adv: rx: adv type 0
d_01: @00:00:01.087228  [00:00:01.087,219] <inf> test_adv: rx: adv test +112 ms
d_01: @00:00:01.194223  [00:00:01.194,213] <inf> test_adv: rx: adv type 0
d_01: @00:00:01.194223  [00:00:01.194,213] <inf> test_adv: rx: adv test +109 ms
d_01: @00:00:01.299295  [00:00:01.299,285] <inf> test_adv: rx: adv type 0
d_01: @00:00:01.299295  [00:00:01.299,285] <inf> test_adv: rx: adv test +108 ms
d_01: @00:00:01.400613  [00:00:01.400,604] <inf> test_adv: rx: adv type 0
d_01: @00:00:01.400613  [00:00:01.400,604] <inf> test_adv: rx: adv test +104 ms
d_01: @00:00:01.506326  [00:00:01.506,317] <inf> test_adv: rx: adv type 0
d_01: @00:00:01.506326  [00:00:01.506,317] <inf> test_adv: rx: adv test +108 ms
d_01: @00:00:01.506326  [00:00:01.506,317] <inf> test_adv: rx completed. stop observer.
d_00: @00:00:01.765625  [00:00:01.765,625] <inf> test_adv: Provision device under test
d_00: @00:00:01.765625  [00:00:01.765,625] <inf> bt_mesh_main: Primary Element: 0x0001
d_00: @00:00:01.765625  [00:00:01.765,625] <dbg> bt_mesh_main: bt_mesh_provision: net_idx 0x0000 flags 0x00 iv_index 0x0000
d_01: @00:00:02.769138  [00:00:02.769,134] <inf> test_adv: rx: adv type 0
d_01: @00:00:02.769138  [00:00:02.769,134] <inf> test_adv: rx: adv test +0 ms
d_01: @00:00:03.771777  [00:00:03.771,759] <inf> test_adv: rx: adv type 0
d_01: @00:00:03.771777  [00:00:03.771,759] <inf> test_adv: rx: adv test +1027 ms
d_01: @00:00:04.778169  [00:00:04.778,167] <inf> test_adv: rx: adv type 0
d_01: @00:00:04.778169  [00:00:04.778,167] <inf> test_adv: rx: adv test +1030 ms
d_01: @00:00:05.779514  [00:00:05.779,510] <inf> test_adv: rx: adv type 0
d_01: @00:00:05.779514  [00:00:05.779,510] <inf> test_adv: rx: adv test +1026 ms
d_01: @00:00:06.787524  [00:00:06.787,506] <inf> test_adv: rx: adv type 0
d_01: @00:00:06.787524  [00:00:06.787,506] <inf> test_adv: rx: adv test +1032 ms
d_01: @00:00:06.787524  [00:00:06.787,506] <inf> test_adv: rx completed. stop observer.
d_01: @00:00:07.629173  [00:00:07.629,150] <inf> test_adv: rx: adv test +0 ms
d_01: @00:00:07.659141  [00:00:07.659,118] <inf> test_adv: rx: adv test +30 ms
d_01: @00:00:07.685936  [00:00:07.685,913] <inf> test_adv: rx: adv test +28 ms
d_01: @00:00:07.705955  [00:00:07.705,932] <inf> test_adv: rx: adv test +20 ms
d_01: @00:00:07.734215  [00:00:07.734,191] <inf> test_adv: rx: adv test +29 ms
d_01: @00:00:07.763939  [00:00:07.763,916] <inf> test_adv: rx: adv test +31 ms
d_01: @00:00:07.763939  [00:00:07.763,916] <inf> test_adv: rx completed. stop observer.
d_01: @00:00:07.786746  [00:00:07.786,743] <inf> test_adv: rx: adv type 0
d_01: @00:00:07.786746  [00:00:07.786,743] <inf> test_adv: rx: adv test +0 ms
d_01: @00:00:08.791125  [00:00:08.791,107] <inf> test_adv: rx: adv type 0
d_01: @00:00:08.791125  [00:00:08.791,107] <inf> test_adv: rx: adv test +1029 ms
d_01: @00:00:09.795654  [00:00:09.795,623] <inf> test_adv: rx: adv type 0
d_01: @00:00:09.795654  [00:00:09.795,623] <inf> test_adv: rx: adv test +1028 ms
d_01: @00:00:10.795536  [00:00:10.795,532] <inf> test_adv: rx: adv type 0
d_01: @00:00:10.795536  [00:00:10.795,532] <inf> test_adv: rx: adv test +1024 ms
d_01: @00:00:10.795536  [00:00:10.795,532] <inf> test_adv: rx completed. stop observer.
d_01: @00:00:10.795537 INFO: test_rx_proxy_mixin PASSED
d_00: @00:00:13.633790 INFO: test_tx_proxy_mixin PASSED
```

**Review by Balaklaka — DISMISSED** [review]

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c
Comment (PavelVPV) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c]: It looks like a rounding issue, because if set `CONFIG_SYS_CLOCK_TICKS_PER_SEC=32768` the so called tolerance will be within 10ms which is closer to the truth. Maybe interval + 15ms is reasonable fluctuation for this kind of test. Delta for interval check could I think also be updated in `xmit_scan_cb`. At least it shouldn't be double of interval.

**Review by PavelVPV — CHANGES_REQUESTED** [review]

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c
Comment (PavelVPV) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c]: This is a Secure Network Beacon. If I disabled it through `CONFIG_BT_MESH_BEACON_ENABLE` and set transmit to `5000 / 1000`, then test passes. I guess the beacon can be disabled through Config Client.

Comment (alxelax) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c]: How did Secure Network Beacon pass Proxy Service parser? 
https://github.com/zephyrproject-rtos/zephyr/pull/50195/files#diff-27be331c9da5ebf00b50fdde2d30791a6fd470e79c5f65fa3fde138dc506a8a5R153

Comment (PavelVPV) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_advertiser.c]: It didn't, it affected time when proxy beacon was sent. I suggest you to try disabling SNB.

**Review by PavelVPV — APPROVED** [review]

**Comment by alxelax** [comment]
@Andrewpini, @Balaklaka, seems I resolved all comments in this PR. Could you reapprove this if there are no objections?

**Review by Balaklaka — APPROVED** [review]

**Review by Andrewpini — APPROVED** [review]

**Pull request merged**
